File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jerry/components/JerryConfig.java
package org.firstinspires.ftc.teamcode.jerry.components;

import com.qualcomm.hardware.bosch.BNO055IMU;
import com.qualcomm.hardware.bosch.JustLoggingAccelerationIntegrator;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import com.qualcomm.robotcore.hardware.TouchSensor;

import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName;
import org.firstinspires.ftc.teamcode.common.Deadwheel;
import org.firstinspires.ftc.teamcode.common.RobotConfig;

import java.util.ArrayList;
import java.util.Iterator;

public class JerryConfig extends RobotConfig {

    // Add declarations here
    public WebcamName webcam;
    public int monitorID;
    public DcMotorEx bl;
    public DcMotorEx br;
    public DcMotorEx fl;
    public DcMotorEx fr;
    public Deadwheel x, y;
    public Servo claw1;
    public Servo claw2;
    public DcMotorEx arm1;
    public DcMotorEx arm2;
    public BNO055IMU imu;
    public TouchSensor limit;

    public static JerryConfig newConfig(HardwareMap hardwareMap, Telemetry telemetry) {
        JerryConfig config = new JerryConfig();
        config.init(hardwareMap, telemetry);
        return config;
    }

    @Override
    protected void init(HardwareMap hardwareMap, Telemetry telemetry) {
        setTelemetry(telemetry);
//        setHardwareMap(hardwareMap);

        webcam = (WebcamName) getHardware("Webcam", WebcamName.class);
        monitorID = hardwareMap.appContext.getResources().getIdentifier(
                "cameraMonitorViewId", "id", hardwareMap.appContext.getPackageName());

        bl = (DcMotorEx) getHardware("Back Left", DcMotorEx.class);
        br = (DcMotorEx) getHardware("Back Right", DcMotorEx.class);
        fl = (DcMotorEx) getHardware("Front Left", DcMotorEx.class);
        fr = (DcMotorEx) getHardware("Front Right", DcMotorEx.class);
        arm1 = (DcMotorEx) getHardware("Arm Motor 1", DcMotorEx.class);
        arm2 = (DcMotorEx) getHardware("Arm Motor 2", DcMotorEx.class);
        claw1 = (Servo) getHardware("Claw Servo 1", Servo.class);
        claw2 = (Servo) getHardware("Claw Servo 2", Servo.class);
        limit = (TouchSensor) getHardware("Arm Stop", TouchSensor.class);

        // Motor direction configuration
        fl.setDirection(DcMotorEx.Direction.REVERSE);
        fr.setDirection(DcMotorEx.Direction.FORWARD);
        bl.setDirection(DcMotorEx.Direction.FORWARD);
        br.setDirection(DcMotorEx.Direction.REVERSE);

        // Encoder configuration (Using modified DcMotor classes with built-in distance calculations)
        // These encoders will mirror a DcMotor, but will be attached to their own port (for example,
        // motor 0 and 1 on Expansion Hub, but without any power connection)
        x = (Deadwheel) getHardware("X Encoder", Deadwheel.class);
        y = (Deadwheel) getHardware("Y Encoder", Deadwheel.class);


        // Control Hub IMU configuration
        // This uses the legacy methods for IMU initialisation, this should be refactored and updated
        // at some point in time. (23 Nov 2022)
        BNO055IMU.Parameters parameters = new BNO055IMU.Parameters();
        parameters.angleUnit = BNO055IMU.AngleUnit.DEGREES;
        parameters.accelUnit = BNO055IMU.AccelUnit.METERS_PERSEC_PERSEC;
        parameters.loggingEnabled = true;
        parameters.loggingTag = "IMU";
        parameters.accelerationIntegrationAlgorithm = new JustLoggingAccelerationIntegrator();

        imu = (BNO055IMU) getHardware("imu", BNO055IMU.class);
        if (imu != null)
            imu.initialize(parameters);

        ArrayList<String> errors = getHardwareErrors();
        if (errors == null) {
            getTelemetry().addData("BunyipsOpMode Status", "ROBOT CONFIGURATION COMPLETED SUCCESSFULLY WITH NO ERRORS.");
            return;
        }

        getTelemetry().addData("BunyipsOpMode Status", "ERROR(S) DURING CONFIGURATION, THESE DEVICES WERE NOT ABLE TO BE CONFIGURED.");
        Iterator<String> error = errors.iterator();
        for (int i = 0; i < errors.size(); i++) {
            getTelemetry().addData(String.valueOf(i), error.next());
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jerry/components/JerryConfig.kt
package org.firstinspires.ftc.teamcode.jerry.components

import com.qualcomm.hardware.rev.RevHubOrientationOnRobot
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction
import com.qualcomm.robotcore.hardware.HardwareDevice
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.IMU
import com.qualcomm.robotcore.hardware.Servo
import com.qualcomm.robotcore.hardware.TouchSensor
import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName
import org.firstinspires.ftc.teamcode.common.RobotConfig
import kotlin.properties.Delegates

/**
 * Jerry robot configuration and hardware declarations.
 */
class JerryConfig(override var hardwareMap: HardwareMap) : RobotConfig() {
    var monitorID by Delegates.notNull<Int>()
    var webcam: WebcamName? = null
    var bl: DcMotorEx? = null
    var br: DcMotorEx? = null
    var fl: DcMotorEx? = null
    var fr: DcMotorEx? = null
    var claw: Servo? = null
    var arm1: DcMotorEx? = null
    var arm2: DcMotorEx? = null
    var imu: IMU? = null
    var limit: TouchSensor? = null

    val driveMotors: List<DcMotorEx?>
        get() = listOf(bl, br, fl, fr)

    val armComponents: List<HardwareDevice?>
        get() = listOf(arm1, arm2, claw, limit)

    override val deviceMappings: HashMap<HardwareDevice?, String> = hashMapOf(
        webcam to "Webcam",
        bl to "Back Left",
        br to "Back Right",
        fl to "Front Left",
        fr to "Front Right",
        arm1 to "Arm Motor 1",
        arm2 to "Arm Motor 2",
        claw to "Claw Servo",
        limit to "Arm Stop",
        imu to "ch_imu"
    )

    override fun init() {
        for ((device, deviceName) in deviceMappings) {
            val hardwareDevice = getHardware(deviceName, device?.javaClass)

            // Assign the hardware device to the corresponding variable
            when (device) {
                webcam -> webcam = hardwareDevice as? WebcamName
                bl -> bl = hardwareDevice as? DcMotorEx
                br -> br = hardwareDevice as? DcMotorEx
                fl -> fl = hardwareDevice as? DcMotorEx
                fr -> fr = hardwareDevice as? DcMotorEx
                arm1 -> arm1 = hardwareDevice as? DcMotorEx
                arm2 -> arm2 = hardwareDevice as? DcMotorEx
                claw -> claw = hardwareDevice as? Servo
                limit -> limit = hardwareDevice as? TouchSensor
                imu -> imu = hardwareDevice as? IMU
            }

            // Update mapping with the proper hardware device
            deviceMappings.remove(device)
            deviceMappings[hardwareDevice] = deviceName
        }

        monitorID = hardwareMap.appContext.resources.getIdentifier(
            "cameraMonitorViewId", "id", hardwareMap.appContext.packageName
        )

        /*
           Deadwheels attached as an encoder on the front two motors
           X encoder = port 3, Front Left
           Y encoder = port 4, Front Right
         */

        // Motor direction configuration
        fl?.direction = Direction.REVERSE
        fr?.direction = Direction.FORWARD
        bl?.direction = Direction.FORWARD
        br?.direction = Direction.REVERSE

        // Control Hub IMU configuration
        val parameters = IMU.Parameters(
            RevHubOrientationOnRobot(
                RevHubOrientationOnRobot.LogoFacingDirection.UP,
                RevHubOrientationOnRobot.UsbFacingDirection.LEFT
            )
        )
        imu?.initialize(parameters)
    }

    companion object {
        fun newConfig(hardwareMap: HardwareMap): JerryConfig {
            val config = JerryConfig(hardwareMap)
            config.init()
            return config
        }
    }
}